### PR TITLE
e-puck2 server

### PR DESCRIPTION
**Pull request by omichel**
This PR adds a sample simulation where an e-puck2 robot runs a TCP/IP server using the same protocol as the one used by the real e-puck2. It makes it possible to use the TCP/IP version of the [EPuckMonitor](http://www.gctronic.com/doc/index.php?title=e-puck2_PC_side_development#WiFi) program to control this simulated robot.

### CODE REVIEW

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
👍

**Comment by DavidMansolino** [comment]
👀

File: projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+  socklen_t asize;
+#else
+  int asize;
+#endif
+  asize = sizeof(struct sockaddr_in);
+  cfd = accept(server_fd, (struct sockaddr *)&client, &asize);
+  if (cfd == -1) {
+#ifdef _WIN32
+    int e = WSAGetLastError();
+    if (e == WSAEWOULDBLOCK)
```
Comment (DavidMansolino) [inline projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c]: It doesn't seem useful to store this value in a variable.

File: projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+#include "play_melody.h"
+
+#define SOCKET_PORT 1000
+#define MOTOR_RATIO 0.00628
+
+static bool socket_init() {
+#ifdef _WIN32 /* initialize the socket API */
+  WSADATA info;
+  int rc = WSAStartup(MAKEWORD(1, 1), &info); /* Winsock 1.1 */
+  if (rc != 0) {
```
Comment (DavidMansolino) [inline projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c]: It doesn't seem useful to store this value in a variable.

File: docs/guide/epuck.md [hunk 3]
```diff
@@ -387,6 +391,13 @@ This is done using the distance sensors of the e-puck.
 
 ![e-puck2.wbt.png](images/robots/epuck/e-puck2.wbt.png) This is the same simulation as the above one, except that it uses an "e-puck2" model.
 
+#### e-puck2\_server.wbt
+
+![e-puck2.wbt.png](images/robots/epuck/e-puck2.wbt.png) This is a simulation of an e-puck2 robot running a controller that implements a TCP/IP server emulating the e-puck2 Wi-Fi protocol.
```
[the remark this answers is not in the document]
Comment (omichel) [reply docs/guide/epuck.md]: I am not sure, that would be too small to be readable...

File: projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+          v = value;
+          sensors_buffer[i * 2 + 54] = *((const unsigned char *)&v);
+          sensors_buffer[i * 2 + 55] = *((const unsigned char *)(&v) + 1);
+        }
+        const double *values = wb_accelerometer_get_values(accelerometer);
+        const double calibration_k[3] = {-9.81 / 760.0, 9.81 / 760.0, 9.81 / 760.0};
+        short int acc[3];
+        for (int i = 0; i < 3; i++) {
+          acc[i] = values[i] / calibration_k[i];
+          acc[i] += 2 - rand() % 5; /* add some random noise */
```
[the remark this answers is not in the document]
Comment (omichel) [reply projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c]: Unfortunately, this is currently not possible from within Webots.

**Comment by fabienrohrer** [comment]
I just tried to build the new controller on macOS and there are several build errors. I can fix them when you're ready.

File: projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c (hunk 1 above)
Comment (DavidMansolino) [reply projects/robots/gctronic/e-puck/controllers/e-puck2_server/e-puck2_server.c]: It seems you have fixed this in 51ad2edd4fcfbac26cd393e5350f4355defa8130 and the re-introduced it in 4fa865ad39ccfe46c471e7506df06e474bbdb63a

**Comment by omichel** [comment]
Gladly, please go ahead.

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by fabienrohrer** [comment]
Travis complains that play_melody.c (and probably play_melody.h) doesn't contain the correct Apache 2.0 License. Should we skip this file in the test or add the license header?

**Comment by omichel** [comment]
I don't know about `play_melody` as I copied the file from the source code of the e-puck and there was no license. Not sure we are allowed to add the Apache 2.0 license here. I would leave it as is (mentioning the original author, and not adding a license).

**Comment by fabienrohrer** [comment]
Ok, then I'm removing it from the test.

**Comment by omichel** [comment]
Thank you.

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
😭

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Comment by fabienrohrer** [comment]
test me

**Pull request merged**